### PR TITLE
Fix FP32 fuse-add path (FPU add to SFPU add) in pre-allgather Welford layernorm

### LINKED ISSUE #45231: Layernorm can produce catastrophically incorrect results with biased FP32 inputs and residual input
When layernorm is called with residual input, pre-add is done through `add_tiles`, whose FPU SrcA reads input data in TF32 (10-bit mantissa) regardless of any precision-preservation plumbing downstream. If input data has a large offset (e.g. 1e06), TF32 ULP at such offset is ~512, so any input values between ~1e06 and ~(1e06 + 512) in cb_in can collapse to a single TF32 value even before reaching multiplication, resulting in variance ~ 0 when significant variance exists in input data.

How severe this issue is depends on the likelihood of data with large offset being an input to layernorm.

When fixing, re-enable xfailing test cases in tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py

### PR DESCRIPTION
**Pull request by ign-abirami**
### Ticket

- #45231

### Problem Description
When LayerNorm runs with residual input enabled, the residual pre-add path uses add_tiles, where FPU SrcA reads data in TF32 precision (10-bit mantissa). For large-offset inputs (e.g. 1e6), TF32 has a coarse ULP (~512), causing nearby values to round to the same representable value during pre-add.
This precision loss happens before downstream high-precision variance computation, collapsing meaningful input variation.

### What's Changed
**1.`layernorm_pre_allgather_welford.cpp`**
- In **Fuse-add (input + residual)** path: changed FPU add_tiles to SFPU copy_tile + add_binary_tile when fp32 unpack is active.
- Welford loop (fuse path): per-tile transpose_wh_init_short + welford_init<PreserveStats>() when fp32 unpack is active (replay-buffer fix after transpose).
- bf16 / non-fp32-unpack path unchanged (still uses add_tiles)

**2.`layernorm_pre_all_gather_welford_program_factory.cpp`**
- welford_unpack_fp32_active enabled for fuse path too
- Made Circular buffer as UnpackToDestFp32 on c_0 (input), c_5 (residual), and c_3 (post-add) when fuse_add enabled + FP32 input.

**3.`test_distributed_layernorm_pre_allgather.py`**
- Removed xfail on use_residual=True in test_layernorm_pre_all_gather_welford_fp32_precision.

### Perf-report (FPU vs SFPU)

| Shape | Offset | Use Residual | OP Code | FPU Device Time (ns) | SPFU Device Time (ns) | Diff (ns) | Diff % |
|--------|--------|--------------|---------|----------------------|-----------------------|------------|--------|
| (1, 1, 32, 128) | 0 | FALSE | LayerNormPreAllGatherDeviceOperation | 7807 | 7791 | 16 | 0.20% |
| (1, 1, 32, 128) | 0 | TRUE | LayerNormPreAllGatherDeviceOperation | 11253 | 13944 | -2691 | -23.91% |
| (1, 1, 32, 128) | 1000000 | FALSE | LayerNormPreAllGatherDeviceOperation | 7790 | 7804 | -14 | -0.18% |
| (1, 1, 32, 128) | 1000000 | TRUE | LayerNormPreAllGatherDeviceOperation | 11260 | 14259 | -2999 | -26.63% |
| (1, 5, 30, 72) | 0 | FALSE | LayerNormPreAllGatherDeviceOperation | 6702 | 6706 | -4 | -0.06% |
| (1, 5, 30, 72) | 0 | FALSE | LayerNormPreAllGatherDeviceOperation | 6569 | 6551 | 18 | 0.27% |
| (1, 5, 30, 72) | 0 | TRUE | LayerNormPreAllGatherDeviceOperation | 9255 | 11475 | -2220 | -23.99% |
| (1, 5, 30, 72) | 0 | TRUE | LayerNormPreAllGatherDeviceOperation | 9198 | 11226 | -2028 | -22.05% |
| (1, 5, 30, 72) | 1000000 | FALSE | LayerNormPreAllGatherDeviceOperation | 6667 | 6627 | 40 | 0.60% |
| (1, 5, 30, 72) | 1000000 | FALSE | LayerNormPreAllGatherDeviceOperation | 6457 | 6467 | -10 | -0.15% |
| (1, 5, 30, 72) | 1000000 | TRUE | LayerNormPreAllGatherDeviceOperation | 9251 | 11260 | -2009 | -21.72% |
| (1, 5, 30, 72) | 1000000 | TRUE | LayerNormPreAllGatherDeviceOperation | 9187 | 11239 | -2052 | -22.34% |
| (4, 1, 64, 128) | 0 | FALSE | LayerNormPreAllGatherDeviceOperation | 8295 | 8215 | 80 | 0.96% |
| (4, 1, 64, 128) | 0 | TRUE | LayerNormPreAllGatherDeviceOperation | 12145 | 15095 | -2950 | -24.29% |
| (4, 1, 64, 128) | 1000000 | FALSE | LayerNormPreAllGatherDeviceOperation | 8180 | 8244 | -64 | -0.78% |
| (4, 1, 64, 128) | 1000000 | TRUE | LayerNormPreAllGatherDeviceOperation | 12085 | 15206 | -3121 | -25.83% |

### Note
- **Mean stagnation** occurs for l**arge offset values** during mean calculation in the Welford kernel due to FP32 precision limitations (ULP increases to ~0.0625 at higher magnitudes). Since variance computation depends on the accumulated mean, this precision loss propagates and leads to variance error accumulation, resulting in **PCC degradation**. This behavior is **independent of residual input** and is consistently observed for both `residual_input = true` and `residual_input = false`.
- Refer below for test results and welford replication results.

## Test Results

**Test:**  
`tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py::test_layernorm_pre_all_gather_welford_fp32_precision`  
**Note:  var_pcc_threshold = 0.95**
 
| Input Shape | Offset | Use Residual | Mean Status | Variance Status | Var PCC |
|------------|--------|--------------|-------------|-----------------|---------|
| (1, 1, 32, 384)  | 1000000 | FALSE | PASSED | FAILED | 0.929945 |
| (1, 1, 32, 384)  | 1000000 | TRUE  | PASSED | FAILED | 0.929945 |
| (1, 1, 32, 512)  | 1000000 | FALSE | PASSED | FAILED | 0.925210 |
| (1, 1, 32, 512)  | 1000000 | TRUE  | PASSED | FAILED | 0.925210 |
| (1, 1, 32, 576)  | 1000000 | FALSE | PASSED | FAILED | 0.928660 |
| (1, 1, 32, 576)  | 1000000 | TRUE  | PASSED | FAILED | 0.928660 |
| (1, 1, 32, 640)  | 1000000 | FALSE | PASSED | FAILED | 0.804922 |
| (1, 1, 32, 640)  | 1000000 | TRUE  | PASSED | FAILED | 0.804922 |
| (1, 1, 32, 704)  | 1000000 | FALSE | PASSED | FAILED | 0.942954 |
| (1, 1, 32, 704)  | 1000000 | TRUE  | PASSED | FAILED | 0.942954 |
| (1, 1, 32, 768)  | 1000000 | FALSE | PASSED | FAILED | 0.923002 |
| (1, 1, 32, 768)  | 1000000 | TRUE  | PASSED | FAILED | 0.923002 |
| (1, 1, 32, 832)  | 1000000 | FALSE | PASSED | FAILED | 0.928021 |
| (1, 1, 32, 832)  | 1000000 | TRUE  | PASSED | FAILED | 0.928021 |
| (1, 1, 32, 896)  | 1000000 | FALSE | PASSED | FAILED | 0.758836 |
| (1, 1, 32, 896)  | 1000000 | TRUE  | PASSED | FAILED | 0.758836 |
| (1, 1, 32, 960)  | 1000000 | FALSE | PASSED | FAILED | 0.809166 |
| (1, 1, 32, 960)  | 1000000 | TRUE  | PASSED | FAILED | 0.809166 |
| (1, 1, 32, 1024) | 1000000 | FALSE | PASSED | FAILED | 0.870284 |
| (1, 1, 32, 1024) | 1000000 | TRUE  | PASSED | FAILED | 0.870284 |

## FP32 Limitation verified using Python Welford Replication

To verify this FP32 precision limitation, similar Welford steps were replicated for the **shape** **(1, 1, 32, 1024)** with **offset = 1e6** in Python for `mean and variance` calculation. The experiment shows the same numerical behavior as observed in TTNN, confirming that the issue originates from **FP32 precision limitations at large offset values**.

### Welford Replication - Mean

| Row | Welford FP32 Mean (W) | FP64 Mean Ref (torch.mean) | \|W-FP64\| Deviation | TTNN Welford FP32 Mean |
|------|------------------------|----------------------------|----------------------|-------------------------|
| 0  | 1000000.188 | 1000000.032 | 0.15564  | 1000000.188 |
| 1  | 999999.8125 | 999999.9938 | 0.181274 | 999999.8125 |
| 2  | 1000000.063 | 999999.9687 | 0.093811 | 1000000 |
| 3  | 999999.8125 | 999999.9664 | 0.15387  | 999999.8125 |
| 4  | 1000000.063 | 1000000.001 | 0.061462 | 1000000.063 |
| 5  | 999999.625 | 999999.9445 | 0.319519 | 999999.5625 |
| 6  | 1000000.25 | 1000000.03 | 0.220459 | 1000000.25 |
| 7  | 1000000 | 999999.999 | 0.000977 | 1000000 |
| 8  | 999999.9375 | 999999.9789 | 0.041382 | 999999.9375 |
| 9  | 999999.9375 | 999999.9892 | 0.051697 | 999999.9375 |
| 10 | 999999.875 | 999999.9249 | 0.049866 | 999999.875 |
| 11 | 999999.8125 | 1000000.022 | 0.209106 | 999999.75 |
| 12 | 999999.75 | 999999.9858 | 0.235779 | 999999.75 |
| 13 | 999999.9375 | 1000000.017 | 0.079834 | 999999.9375 |
| 14 | 1000000.063 | 999999.9965 | 0.06604 | 1000000.063 |
| 15 | 1000000 | 1000000.03 | 0.030212 | 1000000 |
| 16 | 1000000 | 999999.9776 | 0.0224 | 1000000 |
| 17 | 1000000 | 999999.9512 | 0.048828 | 1000000 |
| 18 | 1000000.25 | 999999.9897 | 0.260254 | 1000000.25 |
| 19 | 999999.9375 | 999999.9421 | 0.004639 | 999999.9375 |
| 20 | 1000000.188 | 999999.9955 | 0.192017 | 1000000.188 |
| 21 | 999999.8125 | 999999.9996 | 0.187073 | 999999.8125 |
| 22 | 1000000.063 | 1000000.03 | 0.032104 | 1000000.125 |
| 23 | 999999.75 | 1000000.029 | 0.279358 | 999999.75 |
| 24 | 999999.8125 | 999999.9656 | 0.153137 | 999999.8125 |
| 25 | 1000000.063 | 999999.9488 | 0.113708 | 1000000.063 |
| 26 | 1000000.125 | 999999.9677 | 0.157288 | 1000000.125 |
| 27 | 1000000.25 | 999999.9462 | 0.303833 | 1000000.25 |
| 28 | 999999.9375 | 999999.9307 | 0.006775 | 999999.9375 |
| 29 | 1000000 | 1000000.007 | 0.006592 | 999999.9375 |
| 30 | 1000000 | 1000000.021 | 0.020508 | 1000000 |
| 31 | 999999.6875 | 1000000.015 | 0.327759 | 999999.6875 |

### Welford Replication — Variance Comparison

| Row | Welford Variance (V) | FP64 Variance Ref (`torch.var`) | \|V - FP64\| Deviation | TTNN Welford Variance |
|------|----------------------|---------------------------------|------------------------|-----------------------|
| 0  | 1.079624176 | 1.056503 | 0.023121 | 1.079624176 |
| 1  | 1.053043365 | 1.021118 | 0.031926 | 1.053421021 |
| 2  | 0.945835114 | 0.936966 | 0.008869 | 0.938907623 |
| 3  | 0.959018707 | 0.934328 | 0.024691 | 0.959018707 |
| 4  | 1.000488281 | 0.993800 | 0.006688 | 1.000488281 |
| 5  | 1.184204102 | 1.080071 | 0.104133 | 1.226741791 |
| 6  | 1.051025391 | 1.002545 | 0.048480 | 1.051025391 |
| 7  | 1.002117157 | 1.002295 | 0.000178 | 1.001949310 |
| 8  | 1.032112122 | 1.030545 | 0.001567 | 1.032112122 |
| 9  | 0.938587189 | 0.935732 | 0.002856 | 0.938587189 |
| 10 | 1.007045746 | 1.004765 | 0.002281 | 1.007022858 |
| 11 | 1.112564087 | 1.069884 | 0.042680 | 1.142005920 |
| 12 | 1.109966278 | 1.053657 | 0.056309 | 1.109966278 |
| 13 | 1.069297791 | 1.062467 | 0.006831 | 1.069026947 |
| 14 | 0.950500488 | 0.947673 | 0.002828 | 0.950500488 |
| 15 | 1.000347137 | 0.999373 | 0.000974 | 0.999832153 |
| 16 | 1.011714935 | 1.011793 | 0.000078 | 1.011013031 |
| 17 | 1.039279938 | 1.037769 | 0.001511 | 1.039279938 |
| 18 | 1.079402924 | 1.010660 | 0.068743 | 1.079402924 |
| 19 | 0.953636169 | 0.952760 | 0.000876 | 0.953140259 |
| 20 | 0.986377716 | 0.949504 | 0.036874 | 0.986377716 |
| 21 | 1.011386871 | 0.976009 | 0.035378 | 1.011386871 |
| 22 | 0.984100342 | 0.982917 | 0.001183 | 0.991550446 |
| 23 | 1.117027283 | 1.040882 | 0.076145 | 1.116874695 |
| 24 | 0.993057251 | 0.969213 | 0.023844 | 0.993057251 |
| 25 | 1.044605255 | 1.032530 | 0.012075 | 1.044605255 |
| 26 | 1.027927399 | 1.001044 | 0.026883 | 1.027927399 |
| 27 | 1.071846008 | 0.977868 | 0.093978 | 1.071846008 |
| 28 | 0.941421509 | 0.942211 | 0.000790 | 0.941421509 |
| 29 | 1.009067535 | 1.009699 | 0.000632 | 1.014041901 |
| 30 | 0.993480682 | 0.993194 | 0.000287 | 0.993480682 |
| 31 | 1.151145935 | 1.043300 | 0.107845 | 1.151145935 |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/fix_layernorm_pre_allgather_welford_45231)

### CODE REVIEW

File: tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py [hunk 1]
```diff
@@ -942,10 +942,6 @@ def test_layernorm_pre_all_gather_welford_fp32_precision(device, inp_shape, offs
     ULP (~512) dwarfs the underlying randn variation (~1), so the variance signal is destroyed
     inside the add before welford runs.
```
[the remark this answers is not in the document]
Comment (ign-abirami) [reply tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py:943]: Resolved.

File: tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py [hunk 2]
```diff
@@ -697,7 +697,7 @@ def test_residual_logical_shape_mismatch_rejected(device, op_name, inp_shape, re
         tt_memory_config=dram_memcfg,
     )
 
-    with pytest.raises(RuntimeError):
+    with pytest.raises(RuntimeError):  # allow-pytest.raises: unchanged from main
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py]: This is not helpful. Remove comment.

**Review by bbradelTT — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py (hunk 2 above)
Comment (ign-abirami) [reply tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py]: Removed in latest push.

Comment (bbradelTT) [reply tests/ttnn/nightly/unit_tests/operations/fused/test_distributed_layernorm_pre_allgather.py]: Thanks.

**Comment by iwroszTT** [comment]
Please fix the test and address the co-pilot's comment

**Comment by ign-msati** [comment]
> Please fix the test and address the co-pilot's comment

addressed

**Review by iwroszTT — APPROVED** [review]
